### PR TITLE
[src] fixed an issue of building with cuda11

### PR DESCRIPTION
**Pull request by jimbozhang**
For the issue #4303 
Tested with both cuda-10.1 and cuda-11.0.

### CODE REVIEW

File: src/makefiles/cuda_64bit.mk [hunk 1]
```diff
@@ -10,7 +10,7 @@ CXXFLAGS += -DHAVE_CUDA -I$(CUDATKDIR)/include -fPIC -pthread -isystem $(OPENFST
 CUDA_INCLUDE= -I$(CUDATKDIR)/include -I$(CUBROOT) -I.. -isystem $(OPENFSTINC)
 CUDA_FLAGS = --compiler-options -fPIC --machine 64 -DHAVE_CUDA \
              -ccbin $(CXX) -DKALDI_DOUBLEPRECISION=$(DOUBLE_PRECISION) \
-             -std=c++14 -DCUDA_API_PER_THREAD_DEFAULT_STREAM -lineinfo \
+             -std=c++11 -DCUDA_API_PER_THREAD_DEFAULT_STREAM -lineinfo \
```
Comment (danpovey) [inline src/makefiles/cuda_64bit.mk:13]: is this change supposed to be here?

File: src/cudamatrix/cublas-wrappers.h [hunk 2]
```diff
@@ -278,25 +278,14 @@ inline cusparseStatus_t cusparse_csrmm2(cusparseHandle_t handle,
   if (status != CUSPARSE_STATUS_SUCCESS) return status;
 
   size_t buffer_size;
```
Comment (danpovey) [inline src/cudamatrix/cublas-wrappers.h:280]: This seeems to be just removing the original support for ALG2... and you said you were compiling with 10.2; in this case surely CUDA_VERSION should not be 11000?  Maybe you changed the installation in some default place but failed to reconfigure?

**Comment by dophist** [comment]
@jimbozhang, for reference, it seems you are reverting the code change in https://github.com/kaldi-asr/kaldi/pull/4304 which fixed the issue in https://github.com/kaldi-asr/kaldi/issues/4303

File: src/makefiles/cuda_64bit.mk (hunk 1 above)
Comment (jimbozhang) [reply src/makefiles/cuda_64bit.mk:13]: With the `devtoolset-7-gcc` of the CentOS Developer Toolset, when I uses `c++14`, the building fails:
https://paste.ubuntu.com/p/N2xWXZ2jtm/

I haven't found why, but using `c++11` could address this problem. :flushed:

**Comment by danpovey** [comment]
Seems to be an issue with certain versions of CUDA 10.1 combined with certain versions of gcc.
I think easiest fix would be to upgrade CUDA to a more recent version of 10.1.
That change would break the Kaldi build in other respects, so we can't do that.
  
https://forums.developer.nvidia.com/t/cuda-10-1-nvidia-youre-now-fixing-gcc-bugs-that-gcc-doesnt-even-have/71063/10

Let's focus on the errors you got with CUDA 11.

**Comment by jimbozhang** [comment]
I compiled it on CUDA 11.0, and I found there is no `CUSPARSE_SPMM_CSR_ALG2 ` in `cusparse.h`:

```bash
$ grep -n CUSPARSE_MM_ALG_DEFAULT /usr/local/cuda/include/cusparse.h
5846:    CUSPARSE_MM_ALG_DEFAULT = 0,

$ grep -n CUSPARSE_SPMM_CSR_ALG2 /usr/local/cuda/include/cusparse.h
(nothing)
```

And the version information of `cusparse.h` is:
```c++
//##############################################################################
//# CUSPARSE VERSION INFORMATION
//##############################################################################

#define CUSPARSE_VER_MAJOR 11
#define CUSPARSE_VER_MINOR 0
#define CUSPARSE_VER_PATCH 0
#define CUSPARSE_VER_BUILD 191
#define CUSPARSE_VERSION (CUSPARSE_VER_MAJOR * 1000 + \
                          CUSPARSE_VER_MINOR *  100 + \
                          CUSPARSE_VER_PATCH)

```

But the CUDA version seems right ...
```bash
$ grep CUDA_VERSION /usr/local/cuda/targets/x86_64-linux/include/cuda.h
#define CUDA_VERSION 11000
```

**Comment by danpovey** [comment]
According to this
https://docs.nvidia.com/cuda/archive/11.0/cusparse/index.html
it should be supported.
You are showing 2 different directories there.  I think there must be some
kind of mixup.

**Comment by jimbozhang** [comment]
Yes, I have noticed the document. Thanks dan, and I'll recheck my CUDA installation.

**Pull request closed without merging**